src/sieve.py

```diff
@@ -1,20 +1,38 @@
-"""Computing primes."""
-
-
-def sieve(n: int) -> list[int]:
-    """
-    Compute all the primes up to (and possibly including) `n`.
-
-    `n` must be positive.
-
-    >>> sieve(15)
-    [2, 3, 5, 7, 11, 13]
-
-    """
-    assert n > 0
-    candidates = list(range(2, n + 1))
-    primes = []
-
-    # FIXME: fill out this bit
-
-    return primes
+"""Computing primes."""
+
+
+def sieve(n: int) -> list[int]:
+    """
+    Compute all the primes up to (and possibly including) `n`.
+
+    `n` must be positive.
+
+    >>> sieve(15)
+    [2, 3, 5, 7, 11, 13]
+
+    """
+    assert n > 0
+    candidates = list(range(2, n + 1))
+    primes = []
+
+    # FIXME: fill out this bit
+
+    # first candidate is prime
+    prime = candidates.pop(0)
+    primes.append(prime)
+
+    # invariant
+    n = len(candidates)
+
+    while n > 0:
+        p = candidates.pop(0)
+        prime_check = [p % prime for prime in primes]
+        if 0 not in prime_check:
+            primes.append(p)
+        else:
+            pass
+
+        # loop breaks if n = 0
+        n = len(candidates)

```

Comment:
If you want to run until `candidates` are empty, you could just use

```python
while candiates:
    ...
```

If you use a list as a truth-value, then an empty list is `False`, and a non-empty list is `True`.

With `if`-statements you don't always need an `else` part. And an `else` part that just `pass` is the same as not having one.

You could thus simplify your solution to something like the one below, were I don't `pop` from `candidates` but just run through them:

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        residuals = [candidate % p for p in primes]
        if 0 not in residuals:
            primes.append(candidate)
    return primes
```

Notice, though, that we compute the residuals for all primes first and then check if they are zero after. We know that `p` shouldn't be added to `primes` the first time we see a `p % prime == 0` so we could also just break there. Then, however, we need to know if we broke out of the loop because we got all the way through or because we explicitly broke out of it. In Python, you can have an `else` part to a loop that is only run if you complete the loop without breaking. Using that, we could end up with this solution:

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        for p in primes:
            if candidate % p == 0:
                break
        else:
            # This only runs if we didn't break
            primes.append(candidate)
    return primes
```